symmetric.go

```diff
@@ -0,0 +1,333 @@
+var Ciphers = map[int]*SymmetricCipher{
+	pkcs11.CKK_AES:            &CipherAES,
+	pkcs11.CKK_DES3:           &CipherDES3,
+	pkcs11.CKK_GENERIC_SECRET: &CipherGeneric,
+	pkcs11.CKK_SHA_1_HMAC:     &CipherHMACSHA1,
+	pkcs11.CKK_SHA224_HMAC:    &CipherHMACSHA224,
+	pkcs11.CKK_SHA256_HMAC:    &CipherHMACSHA256,
+	pkcs11.CKK_SHA384_HMAC:    &CipherHMACSHA384,
+	pkcs11.CKK_SHA512_HMAC:    &CipherHMACSHA512,
+}
+
+// PKCS11SecretKey contains a reference to a loaded PKCS#11 symmetric key object.
+//
+// A *PKCS11SecretKey implements the cipher.Block interface, allowing it be used
+// as the argument to cipher.NewCBCEncrypter and similar methods.
+// For bulk operation this is very inefficient;
+// using NewCBCEncrypterCloser, NewCBCEncrypter or NewCBC from this package is
+// much faster.
+type PKCS11SecretKey struct {
+	PKCS11Object
+
+	// Symmetric cipher information
+	Cipher *SymmetricCipher
+}
+
+// Key generation -------------------------------------------------------------
+
+// GenerateSecretKey creates an secret key of given length and type.
+//
+// The key will have a random label and ID.
+func GenerateSecretKey(bits int, cipher *SymmetricCipher) (*PKCS11SecretKey, error) {
+	return GenerateSecretKeyOnSlot(defaultSlot, nil, nil, bits, cipher)
+}
+
+// GenerateSecretKeyOnSlot creates as symmetric key on a specified slot
+//
+// Either or both label and/or id can be nil, in which case random values will be generated.
+func GenerateSecretKeyOnSlot(slot uint, id []byte, label []byte, bits int, cipher *SymmetricCipher) (*PKCS11SecretKey, error) {
+	var k *PKCS11SecretKey
+	var err error
+	if err = ensureSessions(libHandle, slot); err != nil {
+		return nil, err
+	}
+	err = withSession(slot, func(session *PKCS11Session) error {
+		k, err = GenerateSecretKeyOnSession(session, slot, id, label, bits, cipher)
+		return err
+	})
+	return k, err
+}
+
+// GenerateSecretKeyOnSession creates a symmetric key of given type and
+// length, on a specified session.
+//
+// Either or both label and/or id can be nil, in which case random values will be generated.
+func GenerateSecretKeyOnSession(session *PKCS11Session, slot uint, id []byte, label []byte, bits int, cipher *SymmetricCipher) (key *PKCS11SecretKey, err error) {
+	// TODO refactor with the other key generation implementations
+	if label == nil {
+		if label, err = generateKeyLabel(); err != nil {
+			return nil, err
+		}
+	}
+	if id == nil {
+		if id, err = generateKeyLabel(); err != nil {
+			return nil, err
+		}
+	}
+	var privHandle pkcs11.ObjectHandle
+	// CKK_*_HMAC exists but there is no specific corresponding CKM_*_KEY_GEN
+	// mechanism. Therefore we attempt both CKM_GENERIC_SECRET_KEY_GEN and
+	// vendor-specific mechanisms.
+	for _, genMech := range cipher.GenParams {
+		secretKeyTemplate := []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_SECRET_KEY),
+			pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, genMech.KeyType),
```

Comment:
I've found SoftHSM is occasionally fussy about templates containing (redundant) class and key type definitions. Doesn't appear to affect AES keys, but RSA keys get rejected. Might be a good idea to remove the attributes in case other HSMs/implementations have an issue.

According to the spec it should be fine, but.. you know.

Reply:
Based on the v2.40 spec I agree that these do seem to be optional. However https://github.com/opendnssec/SoftHSMv2/issues/411#issuecomment-412431966 says `CKA_CLASS` and `CKA_KEY_TYPE` are mandatory in SoftHSM. Assuming that's true I think it's a bug but we do need to cope l-)

If we find that there's no one template that satisfies all providers we can start adding special cases or environment variables or something. I'm inclined to say YAGNI for the time being though.